### PR TITLE
Improvements to EventHub/ServiceBus error handling (#1784, #1760)

### PR DESCRIPTION
**Pull request by mathewc**
Fixes for the following issues:

- https://github.com/Azure/azure-webjobs-sdk/issues/1760
- https://github.com/Azure/azure-webjobs-sdk/issues/1784

### CODE REVIEW

File: sample/SampleHost/Program.cs [hunk 1]
```diff
@@ -13,24 +14,32 @@ class Program
     {
         static void Main(string[] args)
         {
-            var config = new JobHostConfiguration();
-            config.Queues.VisibilityTimeout = TimeSpan.FromSeconds(15);
-            config.Queues.MaxDequeueCount = 3;
-            config.LoggerFactory = new LoggerFactory().AddConsole();
+            while (true)
```
Comment (mathewc) [inline sample/SampleHost/Program.cs]: Will undo these changes - was using this create/dispose loop to verify my changes. Before the changes, each restart resulted in ServiceBus error logs being written, now the host restarts are all clean.

File: src/Microsoft.Azure.WebJobs.Host/Extensions/LogLevelExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Host
+{
+    internal static class LogLevelExtensions
+    {
+        internal static TraceLevel ToTraceLevel(this LogLevel logLevel)
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Extensions/LogLevelExtensions.cs]: Copied this from the functions runtime codebase

File: src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusExtensionConfig.cs [hunk 3]
```diff
@@ -51,23 +51,15 @@ public void Initialize(ExtensionConfigContext context)
...
-            {
-                Utility.LogExceptionReceivedEvent(e, "MessageReceiver", context.Trace, context.Config.LoggerFactory);
-            };
-
             // get the services we need to construct our binding providers
             INameResolver nameResolver = context.Config.GetService<INameResolver>();
             IExtensionRegistry extensions = context.Config.GetService<IExtensionRegistry>();
 
+            // register the background exception handler
+            var exceptionHandler = MessagingExceptionHandler.Subscribe(Config.MessageOptions, context.Trace, context.Config.LoggerFactory);
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/Config/ServiceBusExtensionConfig.cs]: The key for ServiceBus issue fix was to capture the exception handler and flow it down to the listener, so it can be unsubscribed BEFORE we abort the messaging factory.

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs [hunk 4]
```diff
@@ -444,6 +444,9 @@ void IExtensionConfigProvider.Initialize(ExtensionConfigContext context)
                 .AddConverter<byte[], EventData>(ConvertBytes2EventData)
                 .AddConverter<EventData, byte[]>(ConvertEventData2Bytes);
 
+            // register the background exception handler
+            MessagingExceptionHandler.Subscribe(_options, context.Trace, context.Config.LoggerFactory);
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs]: We don't need to unsubscribe for EventHubs - the way our listener unregisters the EPH shuts down cleanly. I had flowed it down anyways, however we don't want to unregister before we close the EPH (since any errors happening during graceful shutdown we want to log).

File: src/Microsoft.Azure.WebJobs.ServiceBus/MessagingExceptionHandler.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            }
+
+            public override void Unsubscribe()
+            {
+                _options.ExceptionReceived -= Handle;
+            }
+
+            protected override LogLevel GetLogLevel(Exception ex)
+            {
+                if (ex is ReceiverDisconnectedException ||
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/MessagingExceptionHandler.cs]: This is the key change for the EventHub issue fix - overriding the base level determination to make these Info level rather than errors as they are now (generating CRIs)

File: src/Microsoft.Azure.WebJobs.ServiceBus/MessagingExceptionHandler.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            }
+        }
+
+        private class EventHubExceptionHandler : MessagingExceptionHandler
+        {
+            private readonly EventProcessorOptions _options;
+
+            public EventHubExceptionHandler(EventProcessorOptions options, TraceWriter traceWriter, ILoggerFactory loggerFactory = null) : base("EventProcessorHost", traceWriter, loggerFactory)
+            {
+                _options = options;
```
Comment (paulbatum) [inline src/Microsoft.Azure.WebJobs.ServiceBus/MessagingExceptionHandler.cs]: Null check the options? Otherwise your subscribe/unsubscribe will NRE.

**Review by paulbatum — APPROVED** [review]

**Review by brettsam — APPROVED** [review]

File: src/Microsoft.Azure.WebJobs.ServiceBus/MessagingExceptionHandler.cs (hunk 6 above)
Comment (mathewc) [reply src/Microsoft.Azure.WebJobs.ServiceBus/MessagingExceptionHandler.cs]: Done
